fedora_messaging/schema_utils.py

```diff
@@ -0,0 +1,90 @@
+# Copyright (C) 2018  Red Hat, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+"""
+Utilities for message schemas.
+"""
+
+import collections
+from hashlib import sha256, md5
+
+from six.moves.urllib import parse
+
+try:
+    import libravatar
+except ImportError:
+    # We can do without this dependency.
+    libravatar = None
+
+
+# https://github.com/fedora-infra/fedmsg_meta_fedora_infrastructure/issues/320
+HARDCODED_AVATARS = {
+    "bodhi": "https://apps.fedoraproject.org/img/icons/bodhi-{size}.png",
+    "koschei": "https://apps.fedoraproject.org/img/icons/koschei-{size}.png",
+    # Taskotron may have a new logo at some point.  Check this out:
+    # https://mashaleonova.wordpress.com/2015/08/18/a-logo-for-taskotron/
+    # Ask tflink before actually putting this in place though.  we need
+    # a nice small square version.  It'll look great!
+    # In the meantime, we can use this temporary logo.
+    "taskotron": "https://apps.fedoraproject.org/img/icons/taskotron-{size}.png",
+}
+
+
+def avatar_url(username, size=64, default="retro"):
+    """Get the avatar URL of the provided username.
+
+    Args:
+        username (str): The username to get the avatar of.
+        size (int): Size of the avatar in pixels (it's a square).
+        default (str): Default avatar to return if not found.
+    Returns:
+        str: The URL to the avatar image.
+    """
+    try:
+        return HARDCODED_AVATARS[username].format(size=size)
+    except KeyError:
+        openid = "http://{}.id.fedoraproject.org/".format(username)
+        return libravatar_url(openid=openid, size=size, default=default)
+
+
+def libravatar_url(email=None, openid=None, size=64, default="retro", dns=False):
+    """Get the user's avatar from libravatar.
+
+    Either the user's email or openid must be provided.
+
+    Args:
+        email (str): the user's email
+        openid (str): the user's OpenID
+        size (int): Size of the avatar in pixels (it's a square).
+        default (str): Default avatar to return if not found.
+        dns (bool): Send a DNS SRV query to use Libravatar federation.
+    Returns:
+        str: The URL to the avatar image.
+    """
+    if dns and libravatar is not None:
```

Comment:
This makes the API a bit confusing in my opinion. ``dns=True`` implies using libravatar, but only if it's installed, in which case ``dns`` is ignored. This means when I call this function with ``dns=True`` I might get federation, or maybe I won't, but I certainly wont know without inspecting reading the implementation and checking the returned string.

Since all it does is wrap the libravatar API call, I would recommend that we simply document that ``libravatar.libravatar_url`` exists and allows for federation in this function's docblock. If a user wants that, they'll import libravatar themselves and call it. Otherwise they can use this.